### PR TITLE
Subtype: avoid false alarm caused by eager `forall_exists_subtype`.

### LINKED ISSUE #40865: subtype error with `KV<:Union{K,Ref{K}}` alternate constraint
The following works as expected:
```Julia
function f(
    x::AbstractDict{KV,Tuple{Int}},
    y::Union{Set{K},Dict{K,V}},
) where {K,V, KV <: Union{K,Tuple{K,V}}}
    return "done"
end

K = Int
x = Dict{K,Tuple{Int}}()
y = Set{K}()

julia> f(x, y)
"done"
```
Now replace `K = Int` with `K = Tuple{Int,Int}` and now it throws `MethodError: no method matching`

### PR DESCRIPTION
**Pull request by N5N3**
Nested `forall_exists_subtype` always earse the remaining `Runion`, which should not be ignored if the tailling subtype fails.
This PR tries to keep the `Runion` if the local subtyping has moderate scale.
close #40865.

Note: this PR is based on #48410. The first commit is a squashed version.

### CODE REVIEW

**Comment by N5N3** [comment]
Looks like we don't need #26691 anymore after this change.

**Comment by N5N3** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48441#issuecomment-1407689010) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c7fd50f_vs_f6f6b1c/report.html).

**Comment by N5N3** [comment]
@nanosoldier `runtests(["GAlgebra", "BenchmarkProfiles", "NonlinearSolve", "TensND", "MathML", "StructuralIdentifiability", "NonconvexIpopt", "DifferentiableTrajectoryOptimization", "SIAN", "MomentClosure", "ModelOrderReduction", "BoundaryValueDiffEq", "CellMLToolkit", "PotentialFlow", "PETLION", "Plots", "DifferentiableStateSpaceModels", "ParameterizedFunctions", "MTH229", "AstroChemistry", "OptimizationMOI", "MPIClusterManagers", "Percival", "HighDimPDE", "JSOSolvers", "ODEProblemLibrary", "Quaternionic", "DataDrivenSparse", "ReversePropagation", "Causal", "QuadraticModels", "BSeries", "JumpProblemLibrary", "EasyModelAnalysis", "BlockSystems", "DelaySSAToolkit", "LSODA", "RigorousInvariantMeasures", "DataDrivenSR", "Silico", "SpinWaveTheory", "LicenseGrabber", "NonconvexPavito", "EquationsSolver", "DickeModel", "PostNewtonian", "AlgebraicPetri", "RootedTrees", "SphericalFunctions", "OptimizationOptimJL", "VoronoiFVM", "DirectTrajectoryOptimization", "SolverBenchmark", "NonconvexJuniper", "UncertaintyQuantification", "Stopping", "Sundials", "NLPModelsIpopt", "SolverTools", "ControlSystemsMTK", "DiffEqProblemLibrary", "DECAES", "GeometricIntegrators", "ClimaAtmos", "DataDrivenDMD", "NablaNet", "GasChem", "LegendrePolynomials", "VoronoiFVMDiffEq", "MathOptSymbolicAD", "ODE", "NestedEnvironments", "PreallocationTools", "DataInterpolations", "ODEConvergenceTester", "DerivableFunctions", "PairPlots", "SparseDiffTools", "GeometricEquations", "DiffEqCallbacks", "DCISolver", "DerivableFunctionsBase", "SolverTest", "SBML", "PartiallySeparableNLPModels", "SDEProblemLibrary", "Evolutionary", "Pluto", "SymPy", "NumericalMethodsforEngineers", "QuantumLattices", "ExpressionTreeForge", "FundamentalsNumericalComputation", "Sophon", "TKTDsimulations", "FractionalCalculus", "SummationByPartsOperators", "DataDrivenDiffEq", "ODEInterfaceDiffEq", "SideKicks", "DelayDiffEq", "StochasticDelayDiffEq", "CaNNOLeS", "ExactODEReduction", "SteadyStateDiffEq", "IRKGaussLegendre", "GeneralizedMonteCarlo", "IterativeLQR", "Collide", "Nonconvex", "Optimization", "FletcherPenaltySolver", "RRTMGP", "SolverCore", "CropRootBox", "JosephsonCircuits", "ExactDiagonalization", "ProbNumDiffEq", "NeuronBuilder", "FiniteStateProjection", "PolynomialGTM", "ReactionNetworkImporters", "ReliabilityOptimization", "DiffEqNoiseProcess", "TableTransforms", "DataDrivenLux", "Mehrotra", "ReachabilityAnalysis"])`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48441#issuecomment-1408375963) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f9b6264_vs_45b7e7a/report.html).

**Comment by N5N3** [comment]
PkgEval looks good this time.

File: src/subtype.c [hunk 1]
```diff
@@ -1444,56 +1421,97 @@ static int is_definite_length_tuple_type(jl_value_t *x)
...
+    jl_typeenv_t *t = log;
+    while (t != NULL) {
+        if (x == (jl_value_t *)t->var)
+            return 0;
+        t = t->prev;
     }
+    jl_typeenv_t newlog = { (jl_tvar_t*)x, NULL, log };
+    jl_varbinding_t *xb = lookup(e, (jl_tvar_t *)x);
+    return obviously_nounion(xb ? xb->lb : ((jl_tvar_t *)x)->lb, e, &newlog) &&
+           obviously_nounion(xb ? xb->ub : ((jl_tvar_t *)x)->ub, e, &newlog);
```
Comment (vtjnash) [inline src/subtype.c:1454]: ub and lb also need to be type-equal (such as egal), or there is possibly many unions between those bounds. For example Union<:T<:Any. Or is this just an attempted estimate of the decision points?

I think it might be best to move this commit to a different PR, since I am fairly confident in the rest of this change, but not as clear on this commit change currently.

Comment (N5N3) [reply src/subtype.c:1454]: > Or is this just an attempted estimate of the decision points?

This is an attempted estimation.

With out this commit we will consume at least 2 `Runion` for each invariant parameter.
And some packages (e.g. `Latexify.jl`) form complex `Tuple` with thousands of `Pairs`. (see the test added.)

Comment (vtjnash) [reply src/subtype.c:1454]: Ah, thanks for the explanation. This sounds good to me now. Maybe we could name it `may_contain_union_decision`?

**Review by vtjnash — APPROVED** [review]

**Comment by staticfloat** [comment]
This appears to have caused a significant loadtime regression in CUDA.  X-ref: https://github.com/JuliaLang/julia/issues/48582

**Comment by N5N3** [comment]
Perhaps we encounter another subtyping hanging here. Will take a look tomorrow.
